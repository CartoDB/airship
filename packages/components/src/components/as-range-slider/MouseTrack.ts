export function handleMouseDown(listeners: MouseListeners) {
  const handleMove = (eventProperties) => {
    listeners.move(eventProperties);
    eventProperties.preventDefault();
    eventProperties.stopPropagation();
  };

  const handleRelease = (eventProperties) => {
    _handleRelease(eventProperties, { move: handleMove, release: handleRelease }, listeners);
    eventProperties.preventDefault();
    eventProperties.stopPropagation();
  };

  document.addEventListener('mousemove', handleMove);
  document.addEventListener('touchmove', handleMove);
  document.addEventListener('mouseup', handleRelease);
  document.addEventListener('touchend', handleRelease);

  document.addEventListener('dragstart', preventAndStop);
}

function _handleRelease(eventProperties: MouseEvent, listeners: MouseListeners, customListeners: MouseListeners) {
  document.removeEventListener('mousemove', listeners.move);
  document.removeEventListener('touchmove', listeners.move);
  document.removeEventListener('mouseup', listeners.release);
  document.removeEventListener('touchend', listeners.release);

  document.removeEventListener('dragstart', preventAndStop);

  if (customListeners.move) {
    customListeners.move(eventProperties);
  }

  if (customListeners.release) {
    customListeners.release(eventProperties);
  }
}

function preventAndStop(event: MouseEvent) {
  event.preventDefault();
  event.stopPropagation();
}

interface MouseListeners {
  move?: (eventProperties: MouseEvent) => void;
  release?: (eventProperties: MouseEvent) => void;
}
