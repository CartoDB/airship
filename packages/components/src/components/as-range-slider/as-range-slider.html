<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <base href="/">
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
    <title>Airship</title>
    <link rel="stylesheet" href="/dist/core/core.css">
    <script src="/build/airship.js"></script>

    <style>
      body {
        margin: 10px;
      }
    </style>
  </head>

  <body>
    <as-range-slider min-value="0" max-value="100"></as-range-slider>

    <script>
      const rangeSlider = document.querySelector('as-range-slider');
      rangeSlider.range = [33, 66];
      rangeSlider.formatValue = (value) => (`${value}â‚¬`);
    </script>
  </body>
</html>
