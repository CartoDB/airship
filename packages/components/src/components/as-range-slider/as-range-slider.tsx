import { Component, Event, EventEmitter, Prop, State, Watch } from '@stencil/core';
import { Thumb } from '../as-range-slider-thumb/as-range-slider-thumb';

@Component({
  shadow: false,
  styleUrl: './as-range-slider.scss',
  tag: 'as-range-slider',
})
export class RangeSlider {
  /**
   * Top limit of the range.
   * You cannot drag your slider beyond this value. By default the value is 10.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public maxValue: number = 10;

  /**
   * Bottom limit of the range.
   * You cannot drag your slider below this value. By default the value is 0.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public minValue: number = 0;

  /**
   * Disables component if truthy
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public disabled: boolean = false;

  /**
   * Increment/decrement step of the slider.
   * You can change the step setting a different number to this property. Defaults to 1.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public step: number = 1;

  /**
   * If this property is set to true, and it has multiple value, you can drag the entire track.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public draggable: boolean = false;

  /**
   * If this property is set to true, and it has multiple value, you can drag the entire track.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public formatValue: (value: number) => void;

  @Event() public change: EventEmitter<number | number[]>;
  @Event() public changeStart: EventEmitter<number | number[]>;
  @Event() public changeEnd: EventEmitter<number | number[]>;

  /**
   * Initial value.
   * By default, the value is 0 or the minValue.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public value: number = 0;

  @State() private thumbs: Thumb[] = [];

  /**
   * Initial value.
   * By default, the value is 0 or the minValue.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public range: number[] = [];

  @Watch('value')
  public validateValue(newValue: number) {
    if (!this._isBetweenValidValues(newValue)) {
      throw new Error(`RangeSlider: Value ${newValue} has to be between maxValue and minValue`);
    }
  }

  @Watch('range')
  public validateRange(newRange: number[]) {
    if (!this._sliderHasRange()) {
      throw new Error(`RangeSlider: Range ${newRange} need two values at most`);
    }

    newRange.map((value) => this.validateValue(value));
  }


  public componentWillLoad() {
    this._validateValues();
    this.thumbs = this._createThumbs();
  }

  public render() {
    if (this.thumbs.length < 1) {
      return;
    }

    return (
      <div class='as-range-slider'>
        <div class='as-range-slider__rail'>
          { this.thumbs.map((thumb) => this.renderThumb(thumb)) }
          { this.renderRangeBar() }
        </div>
      </div>
    );
  }

  private renderThumb(thumb: Thumb) {
    return <as-range-slider-thumb
              value={thumb.value}
              minValue={this.minValue}
              maxValue={this.maxValue}
              onValueUpdate={(event) => this.onValueUpdated(thumb, event.detail)}>
           </as-range-slider-thumb>;
  }

  private renderRangeBar() {
    const firstThumbPercentage = this._getPercentage(this.thumbs[0].value);
    const lastThumbPercentage = this._getPercentage(this.thumbs[this.thumbs.length - 1].value);

    return <as-range-slider-bar
             rangeStartPercentage={firstThumbPercentage}
             rangeEndPercentage={lastThumbPercentage}></as-range-slider-bar>;
  }

  private _getPercentage(value) {
    return ((value - this.minValue) / (this.maxValue - this.minValue)) * 100;
  }

  private _validateValues() {
    if (this.value) {
      this.validateValue(this.value);
      return;
    }

    if (this.range) {
      this.validateRange(this.range);
      return;
    }
  }

  private _createThumbs() {
    const hasRangeValues = this.range.length;

    if (!hasRangeValues) {
      return [{
        value: this._isBetweenValidValues(this.value) ? this.value : this.minValue
      }];
    }

    return this.range.map((value) => ({
      value: this._isBetweenValidValues(value) ? value : this.minValue
    }));
  }

  private _isBetweenValidValues(value: number) {
    return value >= this.minValue && value <= this.maxValue;
  }

  private _sliderHasRange() {
    return this.range.length === 2;
  }

  private onValueUpdated(thumb: Thumb, value: number) {
    const [leftThumb, rightThumb] = this.thumbs;
    const isLeftThumb = leftThumb === thumb;
    const isRightThumb = rightThumb === thumb;

    if (isLeftThumb && ((rightThumb.value - 1) < value)) {
      return;
    }

    if (isRightThumb && ((leftThumb.value + 1) > value)) {
      return;
    }

    thumb.value = value;
    this.thumbs = [...this.thumbs];
  }
}
