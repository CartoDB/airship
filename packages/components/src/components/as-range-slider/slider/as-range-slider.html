<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <base href="/">
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0">
  <title>Airship</title>
  <link rel="stylesheet" href="/dist/core/core.css">
  <script src="/build/airship.js"></script>

  <style>
    body {
      margin: 10px;
    }

    ul {
      margin: 0;
      padding: 0;
      list-style-type: none;
    }

    as-range-slider {
      margin-bottom: 50px;
      padding-top: 10px;
    }
  </style>
</head>

<body>
  <ul>
    <li>
      <label>Default</label>
      <as-range-slider id="simple"></as-range-slider>
    </li>
    <li>
      <label>Default disabled</label>
      <as-range-slider disabled="true"></as-range-slider>
    </li>
    <li>
      <label>With multivalue</label>
      <as-range-slider id="range" min-value="25" max-value="75"></as-range-slider>
    </li>
  </ul>


  <script>
    // simple
    const simpleRangeSlider = document.querySelector('#simple');
    simpleRangeSlider.value = 5;
    simpleRangeSlider.addEventListener('change', (event) => console.log('Simple changed:', event.detail));

    // range
    const rangeSlider = document.querySelector('#range');
    rangeSlider.range = [30, 70];
    rangeSlider.step = 5;
    rangeSlider.draggable = true;
    rangeSlider.formatValue = (value) => (`${value}â‚¬`);
    rangeSlider.addEventListener('change', (event) => console.log('Changed:', event.detail));
    rangeSlider.addEventListener('changeStart', (event) => console.log('Change Start:', event.detail));
    rangeSlider.addEventListener('changeEnd', (event) => console.log('Change End:', event.detail));

  </script>
</body>

</html>
