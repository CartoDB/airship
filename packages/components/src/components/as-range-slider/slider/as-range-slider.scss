@import '../../../../../styles/src/core/_dev';

as-range-slider {
  --as--range-slider--rail-color: var(--as--color--ui-03, #{$color-ui-03});
  --as--range-slider--range-color: var(--as--color--primary, #{$color-primary});

  display: block;
  height: get-spacing(8);

  .as-range-slider--disabled {
    pointer-events: none;
  }

  .as-range-slider__rail {
    display: flex;
    position: relative;
    width: calc(100% - 12px);
    height: 12px;
    margin: auto;

    &::before {
      content: ' ';
      position: absolute;
      z-index: $z-index-0;
      top: 50%;
      left: 0;
      width: 100%;
      height: 2px;
      transform: translate3d(0, -50%, 0);
      background-color: var(--as--range-slider--rail-color, #{$color-ui-03});
      pointer-events: none;
    }
  }
}
