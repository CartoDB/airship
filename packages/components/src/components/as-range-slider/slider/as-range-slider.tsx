import { Component, Event, EventEmitter, Prop, State, Watch } from '@stencil/core';
import { Thumb } from '../thumb/as-range-slider-thumb';

@Component({
  shadow: false,
  styleUrl: './as-range-slider.scss',
  tag: 'as-range-slider',
})
export class RangeSlider {
  /**
   * Initial value.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public value: number;

  /**
   * Initial range.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public range: number[];

  /**
   * Bottom limit of the range.
   * You cannot drag your slider below this value. By default the value is 0.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public minValue: number = 0;

  /**
   * Top limit of the range.
   * You cannot drag your slider beyond this value. By default the value is 10.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public maxValue: number = 10;

  /**
   * Increment/decrement step of the slider.
   * You can change the step setting a different number to this property. Defaults to 1.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public step: number = 1;

  /**
   * Disables component if truthy
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public disabled: boolean = false;

  /**
   * If this property is set to true, and it has multiple value, you can drag the entire track.
   *
   * @type {number}
   * @memberof RangeSlider
   */
  @Prop() public draggable: boolean = false;

  /**
   * If this property receives a function, it will be used to format the numbers (eg. for adding $ or â‚¬).
   * @type {function (value: number)}
   * @memberof RangeSlider
   */
  @Prop() public formatValue: (value: number) => void;


  @Event() public change: EventEmitter<number[]>;
  @Event() public changeStart: EventEmitter<number[]>;
  @Event() public changeEnd: EventEmitter<number[]>;

  @State() private thumbs: Thumb[] = [];


  @Watch('value')
  public validateValue(newValue: number) {
    if (!this._isBetweenValidValues(newValue)) {
      throw new Error(`RangeSlider: Value ${newValue} has to be between ` +
        `minValue (${this.minValue}) and maxValue (${this.maxValue})`
      );
    }

    this._updateThumbs();
  }

  @Watch('range')
  public validateRange(newRange: number[]) {
    if (newRange.length !== 2) {
      throw new Error(`RangeSlider: Range ${newRange} need two values at most`);
    }

    newRange.map((value) => this.validateValue(value));
    this._updateThumbs();
  }

  public componentWillLoad() {
    this._validateValues();
    this._updateThumbs();
  }

  public render() {
    if (this.thumbs.length < 1) {
      return;
    }

    const cssClasses = {
      'as-range-slider': true,
      'as-range-slider--disabled': this.disabled
    };

    return (
      <div class={cssClasses}>
        <div class='as-range-slider__rail'>
          {this.thumbs.map((thumb) => this._renderThumb(thumb))}
          {this._renderRangeBar()}
        </div>
      </div>
    );
  }

  private _updateThumbs() {
    this.thumbs = this._createThumbs();
  }

  private _renderThumb(thumb: Thumb) {
    return <as-range-slider-thumb
      value={thumb.value}
      valueMin={thumb.valueMin}
      valueMax={thumb.valueMax}
      percentage={thumb.percentage}
      disabled={this.disabled}
      formatValue={this.formatValue}
      onThumbMove={(event) => this._onThumbMove(thumb, event.detail)}
      onThumbIncrease={() => this._onKeyboardThumbMove(thumb, +1)}
      onThumbDecrease={() => this._onKeyboardThumbMove(thumb, -1)}
      onThumbChangeStart={() => this._emitChangeIn(this.changeStart)}
      onThumbChangeEnd={() => this._emitChangeIn(this.changeEnd)}>
    </as-range-slider-thumb>;
  }

  private _renderRangeBar() {
    const [firstThumbPercentage, lastThumbPercentage] = this._getCurrentThumbPercentages();
    const draggable = this.draggable && this.range !== undefined;
    return <as-range-slider-bar
             rangeStartPercentage={firstThumbPercentage}
             rangeEndPercentage={lastThumbPercentage}
             draggable={draggable}
             disabled={this.disabled}
             stepPercentage={this._getStepPercentage()}
             onBarChangeStart={() => this._emitChangeIn(this.changeStart)}
             onBarChangeEnd={() => this._emitChangeIn(this.changeEnd)}
             onBarMove={(event) => this._onBarMove(event)}></as-range-slider-bar>;
  }

  private _getCurrentThumbPercentages() {
    const firstThumbPercentage = this._sliderHasRange() ? this.thumbs[0].percentage : 0;
    const lastThumbPercentage = this.thumbs[this.thumbs.length - 1].percentage;
    return [firstThumbPercentage, lastThumbPercentage];
  }

  private _validateValues() {
    if (this.value) {
      this.validateValue(this.value);
      return;
    }

    if (this.range) {
      this.validateRange(this.range);
      return;
    }
  }

  private _createThumbs() {
    const hasRangeValues = this.range && this.range.length;

    if (!hasRangeValues) {
      return [this._getThumbData(this.value)];
    }

    const thumbs = this.range.map((value) => this._getThumbData(value));
    this._clampThumbValues(thumbs);
    return thumbs;
  }

  private _getThumbData(value) {
    return {
      percentage: this._isBetweenValidValues(value) ?
        this._getPercentage(value)
        : this._getPercentage(this.minValue),
      value: this._isBetweenValidValues(value) ? value : this.minValue,
      valueMax: this.maxValue,
      valueMin: this.minValue
    };
  }

  private _isBetweenValidValues(value: number) {
    return value >= this.minValue && value <= this.maxValue;
  }

  private _sliderHasRange() {
    return this.range && this.range.length === 2;
  }

  private _onKeyboardThumbMove(thumb: Thumb, direction: number) {
    const percentage = this._getPercentage(thumb.value + (direction * this.step));

    if (percentage < 0 || percentage > 100) {
      return;
    }

    this._onThumbMove(thumb, percentage);
  }

  private _onThumbMove(thumb: Thumb, percentage: number) {
    const [leftThumb, rightThumb] = this.thumbs;
    const isLeftThumb = leftThumb === thumb;
    const isRightThumb = rightThumb === thumb;

    const value = this._getValueFromPercentage(percentage);
    const stepValue = this._getStepValue(value);
    const stepPercentage = this._getPercentage(stepValue);

    let valueMin = this.minValue;
    let valueMax = this.maxValue;

    if (this._sliderHasRange() && isLeftThumb) {
      valueMax = (rightThumb.value - this.step);
      if (valueMax < stepValue) {
        return;
      }
    }

    if (this._sliderHasRange() && isRightThumb) {
      valueMin = (leftThumb.value + this.step);
      if (valueMin > stepValue) {
        return;
      }
    }

    thumb.value = stepValue;
    thumb.valueMin = valueMin;
    thumb.valueMax = valueMax;
    thumb.percentage = stepPercentage;

    this.thumbs = [...this.thumbs];

    this._emitChangeIn(this.change);
  }


  private _onBarMove(percentage) {
    const percentageRange = percentage.detail;
    const rangeValues = percentageRange.map((p) => this._getValueFromPercentage(p));
    const stepValues = rangeValues.map((value) => this._getStepValue(value));

    const thumbs = stepValues.map((stepValue) => ({
      percentage: this._getPercentage(stepValue),
      value: stepValue
    }));

    this._clampThumbValues(thumbs);
    this.thumbs = [...thumbs];

    this._emitChangeIn(this.change);
  }

  private _emitChangeIn(eventEmitterInstance: EventEmitter<number | number[]>) {
    const values = this.thumbs.map((thumb) => thumb.value);
    return eventEmitterInstance.emit(values);
  }

  private _getPercentage(value) {
    return ((value - this.minValue) / (this.maxValue - this.minValue)) * 100;
  }

  private _getValueFromPercentage(percentage) {
    return ((percentage * (this.maxValue - this.minValue)) / 100) + this.minValue;
  }

  private _getStepPercentage() {
    const range = (this.maxValue - this.minValue);
    return this.step * 100 / range;
  }

  private _clampThumbValues(thumbs) {
    const [leftThumb, rightThumb] = thumbs;

    leftThumb.valueMin = this.minValue;
    leftThumb.valueMax = Math.min(rightThumb.value - this.step, this.maxValue);

    rightThumb.valueMin = Math.max(this.minValue, leftThumb.value + this.step);
    rightThumb.valueMax = this.maxValue;
  }

  private _getStepValue(value) {
    return Math.round(value / this.step) * this.step;
  }
}
