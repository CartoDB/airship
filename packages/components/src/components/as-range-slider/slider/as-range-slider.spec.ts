import { TestWindow } from '@stencil/core/dist/testing';
import { RangeSlider } from './as-range-slider';

describe('as-range-slider', () => {
  it('should build', () => {
    expect(new RangeSlider()).toBeTruthy();
  });

  describe('Logic', () => {
    let rangeSlider;

    beforeEach(async () => {
      rangeSlider = new RangeSlider();
    });

    it('should have default min and max values', () => {
      expect(rangeSlider.minValue).toEqual(0);
      expect(rangeSlider.maxValue).toEqual(10);
    });

    it('should emit a change event on thumb move', async () => {
      rangeSlider.minValue = 0;
      rangeSlider.maxValue = 1000;
      rangeSlider.range = [500, 600];

      const onChangeSpy = jest.fn();
      rangeSlider.change = { emit: onChangeSpy };

      rangeSlider._updateThumbs(); // >> ctor ?
      const [leftThumb, rightThumb] = rangeSlider.thumbs;

      const maxPercentage = 100;
      rangeSlider._onThumbMove(rightThumb, maxPercentage);
      expect(onChangeSpy).toHaveBeenCalledWith([500, 1000]);

      const minPercentage = 0;
      rangeSlider._onThumbMove(leftThumb, minPercentage);
      expect(onChangeSpy).toHaveBeenCalledWith([0, 1000]);
    });

    it('should emit a change event on bar move', async () => {
      rangeSlider.minValue = 0;
      rangeSlider.maxValue = 1000;
      rangeSlider.range = [500, 600];

      const onChangeSpy = jest.fn();
      rangeSlider.change = { emit: onChangeSpy };

      rangeSlider._updateThumbs();

      rangeSlider._onBarMove({ detail: [40, 50] });
      expect(onChangeSpy).toHaveBeenCalledWith([400, 500]);
    });
  });

  describe('Rendering', () => {
    let element: HTMLAsRangeSliderElement;
    let testWindow: TestWindow;

    beforeEach(async () => {
      testWindow = new TestWindow();
      element = await testWindow.load({
        components: [RangeSlider],
        html: '<as-range-slider></as-range-slider>'
      });
    });

    it('should render without parameters, using defaults', async () => {
      await testWindow.flush();

      expect(element).toMatchSnapshot();
    });

    it('should render 1 thumb + 1 bar when using a value', async () => {
      element.value = 5;
      element.minValue = 0;
      element.maxValue = 10;
      element.step = 1;
      await testWindow.flush();

      expect(element).toMatchSnapshot();
    });

    it('should render 2 thumbs + 1 bar when using a range', async () => {
      element.range = [2, 4];
      element.minValue = 2;
      element.maxValue = 20;
      element.step = 2;
      await testWindow.flush();

      expect(element).toMatchSnapshot();
    });

    it('can be rendered disabled', async () => {
      element.disabled = true;
      await testWindow.flush();

      expect(element).toMatchSnapshot();
    });
  });

  describe('Validation', async () => {
    let element: RangeSlider;

    beforeEach(async () => {
      element = new RangeSlider();
    });

    it('should honor min and max values when setting a value', async () => {
      element.minValue = 50;
      element.maxValue = 100;

      const outOfBounds = () => {
        element.validateValue(2);
      };

      expect(outOfBounds).toThrow(); // TODO this doesn't work, despite of new Error inside a Watch
    });
  });
});
