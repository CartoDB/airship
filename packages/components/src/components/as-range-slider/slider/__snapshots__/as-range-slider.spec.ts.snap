// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-range-slider Rendering can be rendered disabled 1`] = `
<as-range-slider
  class="hydrated"
>
  
  <div
    class="as-range-slider as-range-slider--disabled"
  >
    <div
      class="as-range-slider__rail"
    >
      <as-range-slider-thumb
        disabled=""
        percentage="0"
        value="0"
        valuemax="10"
        valuemin="0"
      />
      <as-range-slider-bar
        disabled=""
        rangeendpercentage="0"
        rangestartpercentage="0"
        steppercentage="10"
      />
    </div>
  </div>
</as-range-slider>
`;

exports[`as-range-slider Rendering should render 1 thumb + 1 bar when using a value 1`] = `
<as-range-slider
  class="hydrated"
>
  
  <div
    class="as-range-slider"
  >
    <div
      class="as-range-slider__rail"
    >
      <as-range-slider-thumb
        percentage="50"
        value="5"
        valuemax="10"
        valuemin="0"
      />
      <as-range-slider-bar
        rangeendpercentage="50"
        rangestartpercentage="0"
        steppercentage="10"
      />
    </div>
  </div>
</as-range-slider>
`;

exports[`as-range-slider Rendering should render 2 thumbs + 1 bar when using a range 1`] = `
<as-range-slider
  class="hydrated"
>
  
  <div
    class="as-range-slider"
  >
    <div
      class="as-range-slider__rail"
    >
      <as-range-slider-thumb
        percentage="20"
        value="2"
        valuemax="3"
        valuemin="0"
      />
      <as-range-slider-thumb
        percentage="40"
        value="4"
        valuemax="10"
        valuemin="3"
      />
      <as-range-slider-bar
        rangeendpercentage="40"
        rangestartpercentage="20"
        steppercentage="11.11111111111111"
      />
    </div>
  </div>
</as-range-slider>
`;

exports[`as-range-slider Rendering should render without parameters, using defaults 1`] = `
<as-range-slider
  class="hydrated"
>
  
  <div
    class="as-range-slider"
  >
    <div
      class="as-range-slider__rail"
    >
      <as-range-slider-thumb
        percentage="0"
        value="0"
        valuemax="10"
        valuemin="0"
      />
      <as-range-slider-bar
        rangeendpercentage="0"
        rangestartpercentage="0"
        steppercentage="10"
      />
    </div>
  </div>
</as-range-slider>
`;
